### ISSUE TITLE
Nullable inference in conditional expressions

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#36096]
```cs
#nullable enable

public class Program
{
    static void M(bool flag, string? x)
    {
        if (flag ? x?.Length == 2 : x?.Length == 3)
        {
            x.ToString(); // warns today, but perhaps shouldn't warn
        }
    }
}
```

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8DEA7AVwwwmAzgAI49TyBYAKEYAEBmC5gJg4EYB2RgG9GFUbwBsHACwUAsgApgAeyUYKAMxIBzJLwAMAfgoAPAJQixwhmJsU06ivM0QtFI8YMA6ADLUtMAAsKAF5gim4QEy9fPH8g0IpWc2tbUStU1ONPABUlAGUYBDRY+VMAbgoAekqKAHcIBDwoChglABMIAE9dYAIYCgAHRACIAeaoAKUiNrwAcn76xotUgF9lijWGTaA===)

Note that if `flag` is changed to the constant `true`, the warning is not given.

The suggestion here is that if either conditional expression arm has a split state, we join the StateWhenTrue on each arm, and join the StateWhenFalse on each arm, and use the resulting split state as the resulting state of the conditional expression.

**Comment by gafter** [comment]
[This behavior is the same as our other flow analyses](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAobAZgAI8KBhCgbxIuasuwBYKBZACmAHt+AGwrB8SUYJHBcASgZMWSgJYA7GBQAeAbkVLmygGYU++CgH5RuCgDIb3HvwCuGzfJBXb93s9ezZevqMxPqhVPgAnDxu2hQA9HEUiAj8CLQAygAM+ABsAKweAKpQcBT8xk6q0FDKAOaqcCgUQvwAxhAiAG4QCMoQwEKlAOSaQ4FKAL6BUyEsAA693fDhOZLCDr4Uan6BwWFaFAC8FPi6s6HYAOwUMAhOcGeTJBNAA=):

``` c#
using System;
public class C {
    public void M(bool b1, bool b2) {
        int x;
        if (b1 ? b2 && M(out x) : b2 && M(out x))
        {
            Console.WriteLine(x); // error CS0165: Use of unassigned local variable 'x'
        }
    }
    private static bool M(out int x)
    {
        x = 1;
        return true;
    }
}
```

**Comment by RikkiGibson** [comment]
I simplified the reproducer and added a description of what the implementation change would probably be.

**Comment by RikkiGibson** [comment]
This will be fixed as part of the "improved definite assignment" feature. #51463
